### DISCUSSION TITLE
Prettifying the output sound file: filtering, noise removal, normalization

### DISCUSSION THREAD

**Discussion by morrowwm** [Nachtzuster/BirdNET-Pi#122] [Ideas]
Hi, I am here from mcguirepr89's repository at the suggestion of @alexbelgium with some work I did as discussed at https://github.com/mcguirepr89/BirdNET-Pi/discussions/253#discussioncomment-9755522

Briefly, I was getting much better sounding output files by reducing noise, filtering frequencies below 1 KHz, and normalizing the volume to a consistent level. Look at the discussion thread cited above for an example. 

Note this is only modifying the output sound file. It has no effect on detection. That's a separate area of research.

This new Nachtzuster fork has some substantial changes, so my previous pull request to mcguirepr89/BirdNET-Pi won't fit. It looks like I could modify this new repository in `scripts/utils/reporting.py` here:
```
def extract(in_file, out_file, start, stop):
```
replace the following sox call with python code to:
* trim the output (as sox currently is doing)
* reduce noise
* highpass filter
* normalize 

```
    result = subprocess.run(['sox', '-V1', f'{in_file}', f'{out_file}', 'trim', f'={start}', f'={stop}'],
                            check=True, capture_output=True)
    ret = result.stdout.decode('utf-8')
    err = result.stderr.decode('utf-8')
    if err:
        raise RuntimeError(f'{ret}:\n {err}')
    return ret
```
Does this look like the right spot?

**Comment by alexbelgium** [comment]
What is nice is that as it is post analysis, it will improve sound of mp3 files without needing people to force gain on input and risking signal saturation

**Comment by lloydbayley** [comment]
If implemented, I believe it should be an optional setting. I would rather hear the exact sound it recorded and analysed.

**Reply by morrowwm** [reply]
I agree. I'll try to get it working at first. 
The original untrimmed `wav` file is available for some time, correct?

**Reply by alexbelgium** [reply]
Not in this fork, it was removed to improve time of analysis and decrease resources usages, although I have reimplemented it in my homeassistant add-on based on this fork

**Reply by lloydbayley** [reply]
What Speedy Gonzalez above said! ^^^^^^^^^ :)

**Reply by lloydbayley** [reply]
@morrowwm Please, do also understand that I am not against the idea, in principal, it's just that I quite like to keep the original intent of the project as the 'prime-directive' as I've seen way too many projects turn into a "What would you like to add today?" kind of thing. Suddenly a robust package is wearing eyeshadow and party-hats. 

I just feel that an option would be a good idea for those data-driven people who don't want to forcibly tidy the recording up. I believe raw and 'pretty' recordings would both have their place. 

The only thing I advocate for is that the raw recordings are sent to BirdWeather for processing. Having said that, ultimately, it is NOT my fork and is in the hands of @Nachtzuster .

**Reply by morrowwm** [reply]
@lloydbayley, totally in agreement with keeping projects slim and svelte. No need for eye-shadow or party hats.

Just out of curiosity, what do you have for a microphone, and recording setup in general?

**Reply by lloydbayley** [reply]
At the risk of disappointing, I am using the second mic I bought for the project which is a cheap and cheerful USB lapel mic [Here](https://www.amazon.com.au/gp/product/B072Q2GH99/). It's not out of stock on that link but I'm sure it could be found as it's a generic thing. Wind shield installed and in a little 3D Printed "House" I made for it up on the roof. The USB cable is extended with [USB over UTP extenders](https://www.amazon.com.au/gp/product/B0029HRYBK) I an use CAT6 instead of an enormously-long USB extension. Nothing too fancy, no tuning needed, I just threw it out there and tested and it works fine so haven't had the need to do any more. :) I also use the old model as the new one detects less than half the birds the old model does here in Australia.

**Comment by morrowwm** [comment]
I've made a preliminary change here: https://github.com/morrowwm/BirdNET-Pi/blame/main/scripts/utils/reporting.py

Only needed to change the one file.

If you want to try it, you will also likely need to install an additional python module:
```
source ~/BirdNET-Pi/birdnet/bin/activate
python3 -m pip install noisereduce
```

**Reply by dalogue1** [reply]
I got the following error when I tried to apply the filter:
(My system is a RPi5 4GB with 1GB NVMe perhaps the filter code is not compatible with RPi5)

> 9:04:32---[server][INFO] 9.0;12.0-('Lonchura punctulata_Scaly-breasted Munia', 0.24030022)
> 19:04:32---[server][INFO] 12.0;15.0-('Human_Human', 0.0)
> 19:04:32---[utils.reporting][INFO] Using python to trim input BirdSongs/StreamData/2024-06-27-birdnet-19:04:17.wav, reduce noise, highpass filter, normalize, then output to BirdSongs/Extracted/By_Date/2024-06-27/Scaly-breasted_Munia/Scaly-breasted_Munia-87-2024-06-27-birdnet-19:04:17.mp3
> 19:04:32---[birdnet_analysis][ERROR] Unexpected error:
> 19:04:32---Traceback (most recent call last):
> 19:04:32---  File "/usr/local/bin/birdnet_analysis.py", line 113, in handle_reporting_queue
> 19:04:32---    detection.file_name_extr = extract_detection(file, detection)
> 19:04:32---                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 19:04:32---  File "BirdNET-Pi/scripts/utils/reporting.py", line 104, in extract_detection
> 19:04:32---    extract_safe(file.file_name, new_file, detection.start, detection.stop)
> 19:04:32---  File "BirdNET-Pi/scripts/utils/reporting.py", line 80, in extract_safe
> 19:04:32---    extract(in_file, out_file, safe_start, safe_stop)
> 19:04:32---  File "BirdNET-Pi/scripts/utils/reporting.py", line 39, in extract
> 19:04:32---    reduced_noise = noisereduce.reduce_noise(y=data, sr=rate)
> 19:04:32---                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 19:04:32---  File "BirdNET-Pi/birdnet/lib/python3.11/site-packages/noisereduce/noisereduce.py", line 186, in reduce_noise
> 19:04:32---    return sg.get_traces()
> 19:04:32---           ^^^^^^^^^^^^^^^
> 19:04:32---  File "BirdNET-Pi/birdnet/lib/python3.11/site-packages/noisereduce/spectralgate/base.py", line 222, in get_traces
> 19:04:32---    filtered_chunk = self.filter_chunk(start_frame=0, end_frame=end_frame)
> 19:04:32---                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 19:04:32---  File "BirdNET-Pi/birdnet/lib/python3.11/site-packages/noisereduce/spectralgate/base.py", line 148, in filter_chunk
> 19:04:32---    padded_chunk = self._read_chunk(i1, i2)
> 19:04:32---                   ^^^^^^^^^^^^^^^^^^^^^^^^
> 19:04:32---  File "BirdNET-Pi/birdnet/lib/python3.11/site-packages/noisereduce/spectralgate/base.py", line 140, in _read_chunk
> 19:04:32---    chunk = np.zeros((self.n_channels, i2 - i1))
> 19:04:32---            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 19:04:32---numpy.core._exceptions._ArrayMemoryError: Unable to allocate 322. GiB for an array with shape (720000, 60002) and data type float64
> 19:04:47---[birdnet_analysis][INFO] Analyzing BirdSongs/StreamData/2024-06-27-birdnet-19:04:32.wav
> 19:04:47---[server][INFO] READING AUDIO DATA...
> 19:04:47---[server][INFO] READING DONE! READ 5 CHUNKS.
>

**Reply by dalogue1** [reply]
I have a wav file that I could test the filter function with at a Terminal command line if I had a python test script to run.

Unfortunately, I am not familiar enough with Python and packages to write the script myself.  However, all the required packages are installed on my system.

Note: I have an A/C unit running almost continually in my backyard that does not interfere with the detections but does make my sound clips not as nice.  I would like to remove that base noise.  I have attached three wav files that have the A/C noise.
[BLUE_JAY_2024-06-23-birdnet-12_44_13.wav.zip](https://github.com/user-attachments/files/16034751/BLUE_JAY_2024-06-23-birdnet-12_44_13.wav.zip)
[MUNIA_2024-06-23-birdnet-19_24_32.wav.zip](https://github.com/user-attachments/files/16034759/MUNIA_2024-06-23-birdnet-19_24_32.wav.zip)
[WHITE_WINGED_DOVE_2024-06-23-birdnet-13_17_43.wav.zip](https://github.com/user-attachments/files/16034762/WHITE_WINGED_DOVE_2024-06-23-birdnet-13_17_43.wav.zip)

**Reply by YvedD** [reply]
following this idea and topic !!!

Grtz,
Yves

**Reply by morrowwm** [reply]
> I have a wav file that I could test the filter function with at a Terminal command line if I had a python test script to run.
> 
> Unfortunately, I am not familiar enough with Python and packages to write the script myself. However, all the required packages are installed on my system.

This script should work standalone.
[reducenoise.py.zip](https://github.com/user-attachments/files/16042322/reducenoise.py.zip)

 If you get this error:
```
ModuleNotFoundError: No module named 'noisereduce'
```
Install the noisereduce module first:
```
python3 -m pip install noisereduce
```
possibly in a virtualenv depending on your usage of python elsewhere.

Then run the test script
```
python3 reducenoise.py BLUE_JAY_2024-06-23-birdnet-12_44_13.wav quiet_blue_jay.wav
```

However, I tried it with one of your sample wav files, and got the error others are reporting:
```
numpy.core._exceptions.MemoryError: Unable to allocate 322. GiB for an array with shape (720000, 60002) and data type float64
```
The python module is trying to allocate 322 GB of memory to hold the data!

Your data is stereo, I suspect that's the problem. I used audacity Tracks/Mix/Mix Stereo to reduce it down to mono, exported a new wav file, ran the test script again, and it completed.
```
python3 reducenoise.py BLUE_JAY_mono.wav quiet_blue_jay.wav
```

The output is distorted quite a bit, unfortunately. There is a pretty severe background noise situation here, though.

[quiet_blue_jay.wav.zip](https://github.com/user-attachments/files/16042333/quiet_blue_jay.wav.zip)

Here are your original input, the mono wav file and the processed output as seen in audacity:
![image](https://github.com/Nachtzuster/BirdNET-Pi/assets/11416685/9a88933e-5c17-469c-82cf-460c74e09c32)


A good reason to make the noise reduction and high pass filter steps optional. We could at least add normalization, it does not change the shape of the sound, only volume level.

I'll do some research on how to handle stereo data. Is there any point in preserving stereo output? I'm assuming no one is actually using a serious stereo recording set up?

**Reply by morrowwm** [reply]
Update, here's a new version of the test script, which will reduce stereo to mono:
[reducenoise.py.zip](https://github.com/user-attachments/files/16042395/reducenoise.py.zip)


As far as I can tell, stereo has two identical channels, so is not really adding anything useful.

**Reply by dalogue1** [reply]
Thank you for the scripts.   I will try them.

My microphone is built by hand and as such only has a single channel.   I suspect the audio capture command is what determines if the audio is stereo or not.  It might be able to be modified.

Maybe your script could convert a stereo file to mono before it tries to do noise reduction.

**Reply by morrowwm** [reply]
_Maybe your script could convert a stereo file to mono before it tries to do noise reduction._

Yes, it does that now in the latest version.

**Reply by YvedD** [reply]
Hi,
How to implement this in my running (test-fase until fall migration starts, by then i will be on full stereo to detect mirating direction also) setup?
Problem for me is that i have no experience on a RPi whatsoever nor linux-like OS

**Reply by YvedD** [reply]
tried this (replaced reportings.py with your version in "filemanager") but i lack "Live Audio" now, any ideas on that?
I have noisereduce installed and reporting OK on my installation.

**Reply by morrowwm** [reply]
> tried this (replaced reportings.py with your version in "filemanager") but i lack "Live Audio" now, any ideas on that? I have noisereduce installed and reporting OK on my installation.

Anything interesting in `/var/log/syslog`?

**Reply by YvedD** [reply]
I finally did get it up and running...but...the sound was totaly not fit to do a good detection, looked phase shifted and pitched and overly loud (normalize!)
The lowest noise level was very well removed...but the rest was to heavy tweaked to use...
Any ideas?

**Reply by morrowwm** [reply]
Can you attach a sample?

I'm looking at how to add configuration. My thoughts now are to only do the normalization for now, and probably make the level configurable.

**Reply by YvedD** [reply]
i think (i'm no sound ing.):
The noise reduction works not to bad at all. someone must invent a proper AI-noise-cancelling for backgroundsound :-)

Problem is the phase shifting and over-normalizing ... if one could keep the correct pitch in the soundclips, and then gradualy set a highpass filter, again without disturbing the pitch and "feel" of the sound, no normalization would be needed, there will always be birds who are to far to detect correctly - as we do in human detectiong in the field ... no?

from next week on i'll have heaps of spare time, as in retired :-) 
I can't program myself (not yet in Phyton 3), if i know how to make a good backup and how to set that back (so i won't lose my current NFC's) i can test things out on a daily basis...

Grtz,
Yves

**Comment by morrowwm** [comment]
@Nachtzuster , what do you think of a reduced scope pull request which just replaces the sox call with "native" python, and normalizes to output to full volume? And perhaps makes the normalization optional?

We can leave out the filtering and noise removal for now.

**Reply by Nachtzuster** [reply]
This would be nice when we got rid of the other issue with sox https://github.com/Nachtzuster/BirdNET-Pi/issues/104
In itself I don't see much value, to be honest

**Comment by alexbelgium** [comment]
Hi, I've tested a new noise reduction script that gives interesting results it seems. I've tested it on several wav files and it gives either equivalent or similar detections levels. Here is how to test it : `sed -i 's|-ac 2 -ar 48000|-ac 2 -ar 48000 -af "highpass=f=250,afftdn=nf=-20"|g' "$HOME"/BirdNET-Pi/scripts/birdnet_recording.sh`
and revert it `sed -i 's|-ac 2 -ar 48000 -af "highpass=f=250,afftdn=nf=-20"|-ac 2 -ar 48000|g' "$HOME"/BirdNET-Pi/scripts/birdnet_recording.sh`

Edit : After a couple hours, the noise reduction allowed a significantly better Signal-Noise ratio (seen from the cleaner spectrogram) for some birds while at least equivalent average detection. There was also more false positives. However, I disabled it for the moment as I prefer the natural sound.
<img src="https://github.com/user-attachments/assets/91f3bde3-6a40-40bd-9aec-b87041df9189" width="300" />

Before / after (same bird singing)
<img src="https://github.com/user-attachments/assets/27854569-0210-4cca-a387-f32d4f96107b" width="300" />

**Reply by dalogue1** [reply]
For people who are curious about the audio filter afftdn options, they are documented here.
https://ffmpeg.org/ffmpeg-filters.html#afftdn

**Comment by jkrauska** [comment]
is this at a point to be made in to an official issue?

**Reply by morrowwm** [reply]
I've come around to Nachtzuster's opinion expressed on Jul 8 above, actually. I don't think it has much value on its own.

**Reply by alexbelgium** [reply]
I've stopped using it too and instead improved my recording system. I've just kept a highband at 250hz for the mp3 recording, but not before the model itself
